tracing/src/main/java/com/palantir/tracing/TraceState.java

```diff
@@ -39,6 +41,8 @@ final class TraceState implements Serializable {
     @Nullable
     private final String forUserAgent;
 
+    private final IdentityHashMap<TraceLocal<?>, Object> traceLocals;
```

Comment:
We can use a ConcurrentHashMap and declare final equals+hashcode on `TraceLocal` which delegate to super (Object) for identity semantics.